idaes/core/util/model_diagnostics.py

```diff
@@ -17,29 +17,987 @@
+
+        uc_vblocks, uc_cblocks = igraph.get_connected_components(uc_var, uc_con)
+        oc_vblocks, oc_cblocks = igraph.get_connected_components(oc_var, oc_con)
+
+        return uc_vblocks, uc_cblocks, oc_vblocks, oc_cblocks
+
+    def display_underconstrained_set(self, stream=stdout):
+        """
+        Prints the variables and constraints in the under-constrained sub-problem
+        from a Dulmage-Mendelsohn partitioning.
+
+        This cane be used to identify the under-defined part of a model and thus
+        where additional information (fixed variables or constraints) are required.
+
+        Args:
+            stream: an I/O object to write the list to (default = stdout)
+
+        Returns:
+            None
+
+        """
+        uc_vblocks, uc_cblocks, _, _ = self.get_dulmage_mendelsohn_partition()
+
+        stream.write("=" * MAX_STR_LENGTH + "\n")
+        stream.write("Dulmage-Mendelsohn Under-Constrained Set\n\n")
+
+        for i, uc_vblock in enumerate(uc_vblocks):
+            stream.write(f"{TAB}Independent Block {i}:\n\n")
+            stream.write(f"{2*TAB}Variables:\n\n")
+            for v in uc_vblock:
+                stream.write(f"{3*TAB}{v.name}\n")
+
+            stream.write(f"\n{2*TAB}Constraints:\n\n")
+            for c in uc_cblocks[i]:
+                stream.write(f"{3*TAB}{c.name}\n")
+            stream.write("\n")
+
+        stream.write("=" * MAX_STR_LENGTH + "\n")
+
+    def display_overconstrained_set(self, stream=stdout):
+        """
+        Prints the variables and constraints in the over-constrained sub-problem
+        from a Dulmage-Mendelsohn partitioning.
+
+        This cane be used to identify the over-defined part of a model and thus
+        where constraints must be removed or variables unfixed.
+
+        Args:
+            stream: an I/O object to write the list to (default = stdout)
+
+        Returns:
+            None
+
+        """
+        _, _, oc_vblocks, oc_cblocks = self.get_dulmage_mendelsohn_partition()
+
+        stream.write("=" * MAX_STR_LENGTH + "\n")
+        stream.write("Dulmage-Mendelsohn Over-Constrained Set\n\n")
+
+        for i, oc_vblock in enumerate(oc_vblocks):
+            stream.write(f"{TAB}Independent Block {i}:\n\n")
+            stream.write(f"{2*TAB}Variables:\n\n")
+            for v in oc_vblock:
+                stream.write(f"{3*TAB}{v.name}\n")
+
+            stream.write(f"\n{2*TAB}Constraints:\n\n")
+            for c in oc_cblocks[i]:
+                stream.write(f"{3*TAB}{c.name}\n")
+            stream.write("\n")
+
+        stream.write("=" * MAX_STR_LENGTH + "\n")
+
+    def display_variables_with_extreme_jacobians(self, stream=stdout):
+        """
+        Prints the variables associated with columns in the Jacobian with extreme
+        L2 norms. This often indicates poorly scaled variables.
+
+        Tolerances can be set via the DiagnosticsToolbox config.
+
```

Comment:
What do you think about allowing tolerances to be set in this function? E.g.
```python
dt = DiagnosticsToolbox(m, jacobian_large_value_caution=1e8)
# Display with 1e8 threshold
dt.display_variables_with_extreme_jacobians(m)
# If the output is too noisy, we may want to display with a higher threshold
# to see only the worst offenders
dt.display_variables_with_extreme_jacobians(m, jacobian_large_value_caution=1e10)
```
This would require something like:
```python
def display_variables_with_extreme_jacobians(self, stream=stdout, **kwds):
    config = self.config(kwds)
    ...
```

I think this is more user-friendly (allows setting tolerances closer to where you actually use them), although the advantage is relatively minor as long as constructing another `DiagnosticsToolbox` is cheap.

Reply:
This goes for all the similar "numerical issues" functions. Leveraging `self.config` would allow these methods to all have consistent call signatures.

Reply:
> This is one case where I intended config blocks to be changed dynamically as required - the methods all look to the config block for the tolerances and there is nothing preventing users from changing when they want to.

If we expect users to change tolerances in the config, I would rather they do it with function arguments than by dynamically altering `self.config`.

> That said, this does run counter to how config blocks are used elsewhere in IDAES where they are only checked once at build time, and thus we discourage changing them after the fact.

But even in IDAES, IIRC initialize methods accept solver arguments at call-time rather than (only) as config arguments. I think allowing this is more intuitive, but I also don't think it is necessary (yet) so have no problem with leaving it out.

Reply:
The old initialization approach used arguments at call time, but the new `Initializer` objects actually work in a similar fashion to this (a config block which users can dynamically update - for better or worse).